pkg/volume/cephfs/cephfs.go

```diff
@@ -0,0 +1,261 @@
+	mon         []string
+	id          string
+	secret      string
+	secret_file string
+	readonly    bool
+	mounter     mount.Interface
+	plugin      *cephfsPlugin
+}
+
+type cephfsBuilder struct {
+	*cephfs
+}
+
+var _ volume.Builder = &cephfsBuilder{}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (cephfsVolume *cephfsBuilder) SetUp() error {
+	return cephfsVolume.SetUpAt(cephfsVolume.GetPath())
+}
+
+func (cephfsVolume *cephfsBuilder) SetUpAt(dir string) error {
+	notMnt, err := cephfsVolume.mounter.IsLikelyNotMountPoint(dir)
+	glog.V(4).Infof("CephFS mount set up: %s %v %v", dir, !notMnt, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if !notMnt {
+		return nil
+	}
+	os.MkdirAll(dir, 0750)
+
+	err = cephfsVolume.execMount(dir)
+	if err == nil {
+		return nil
+	}
+
+	// cleanup upon failure
+	cephfsVolume.cleanup(dir)
+	// return error
+	return err
+}
+
+func (cephfsVolume *cephfsBuilder) IsReadOnly() bool {
+	return cephfsVolume.readonly
+}
+
+type cephfsCleaner struct {
+	*cephfs
+}
+
+var _ volume.Cleaner = &cephfsCleaner{}
+
+func (cephfsVolume *cephfsCleaner) TearDown() error {
+	return cephfsVolume.TearDownAt(cephfsVolume.GetPath())
+}
+
+func (cephfsVolume *cephfsCleaner) TearDownAt(dir string) error {
+	return cephfsVolume.cleanup(dir)
+}
+
+func (cephfsVolume *cephfs) GetPath() string {
+	name := cephfsPluginName
+	return cephfsVolume.plugin.host.GetPodVolumeDir(cephfsVolume.podUID, util.EscapeQualifiedNameForDisk(name), cephfsVolume.volName)
+}
+
+func (cephfsVolume *cephfs) cleanup(dir string) error {
+	noMnt, err := cephfsVolume.mounter.IsLikelyNotMountPoint(dir)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("CephFS: Error checking IsLikelyNotMountPoint: %v", err)
+	}
+	if noMnt {
+		return os.RemoveAll(dir)
+	}
+
+	if err := cephfsVolume.mounter.Unmount(dir); err != nil {
+		return fmt.Errorf("CephFS: Unmounting failed: %v", err)
+	}
+	noMnt, mntErr := cephfsVolume.mounter.IsLikelyNotMountPoint(dir)
+	if mntErr != nil {
+		return fmt.Errorf("CephFS: IsMountpoint check failed: %v", mntErr)
+	}
+	if noMnt {
+		if err := os.RemoveAll(dir); err != nil {
+			return fmt.Errorf("CephFS: removeAll %s/%v", dir, err)
+		}
+	}
+
+	return nil
+}
+
+func (cephfsVolume *cephfs) execMount(mountpoint string) error {
+	// cephfs mount option
+	ceph_opt := ""
+	// override secretfile if secret is provided
+	if cephfsVolume.secret != "" {
+		ceph_opt = "name=" + cephfsVolume.id + ",secret=" + cephfsVolume.secret
+	} else {
+		ceph_opt = "name=" + cephfsVolume.id + ",secretfile=" + cephfsVolume.secret_file
+	}
+	// build option array
+	opt := []string{}
+	if cephfsVolume.readonly {
+		opt = append(opt, "ro")
+	}
+	opt = append(opt, ceph_opt)
+
+	// build src like mon1:6789,mon2:6789,mon3:6789:/
+	hosts := cephfsVolume.mon
+	l := len(hosts)
+	// pass all monitors and let ceph randomize and fail over
+	i := 0
+	src := ""
+	for i = 0; i < l-1; i++ {
+		src += hosts[i] + ","
```

Comment:
right, because we only go to `i-1`, the `i`th host will end the string with `:/`